src/meta-srv/src/procedure/region_migration/upgrade_candidate_region.rs

```diff
@@ -13,24 +13,549 @@
 // limitations under the License.
 
 use std::any::Any;
+use std::time::Duration;
 
+use api::v1::meta::MailboxMessage;
+use common_meta::instruction::{Instruction, InstructionReply, UpgradeRegion, UpgradeRegionReply};
+use common_telemetry::warn;
 use serde::{Deserialize, Serialize};
+use snafu::{ensure, ResultExt};
+use tokio::time::sleep;
 
-use crate::error::Result;
+use super::update_metadata::UpdateMetadata;
+use crate::error::{self, Result};
+use crate::handler::HeartbeatMailbox;
 use crate::procedure::region_migration::{Context, State};
+use crate::service::mailbox::Channel;
+
 #[derive(Debug, Serialize, Deserialize)]
-pub struct UpgradeCandidateRegion;
+pub struct UpgradeCandidateRegion {
+    // The optimistic retry times.
+    optimistic_retry: usize,
+    // The retry initial interval.
+    retry_initial_interval: Duration,
+    // The replay timeout of a instruction.
+    replay_timeout: Duration,
+    // If it's true it requires the candidate region MUST replay the WAL to the latest entry id.
+    // Otherwise, it will rollback to the old leader region.
+    require_ready: bool,
+}
+
+impl Default for UpgradeCandidateRegion {
+    fn default() -> Self {
+        Self {
+            optimistic_retry: 3,
+            retry_initial_interval: Duration::from_millis(500),
+            replay_timeout: Duration::from_millis(1000),
+            require_ready: true,
+        }
+    }
+}
 
 #[async_trait::async_trait]
 #[typetag::serde]
 impl State for UpgradeCandidateRegion {
-    async fn next(&mut self, _ctx: &mut Context) -> Result<Box<dyn State>> {
-        todo!();
+    async fn next(&mut self, ctx: &mut Context) -> Result<Box<dyn State>> {
+        if self.upgrade_region_with_retry(ctx).await {
+            Ok(Box::new(UpdateMetadata::Upgrade))
+        } else {
+            Ok(Box::new(UpdateMetadata::Rollback))
+        }
     }
 
     fn as_any(&self) -> &dyn Any {
         self
     }
 }
 
-impl UpgradeCandidateRegion {}
+impl UpgradeCandidateRegion {
+    const UPGRADE_CANDIDATE_REGION_RTT: Duration = Duration::from_secs(1);
```

Comment:
Move it to https://github.com/GreptimeTeam/greptimedb/blob/develop/src/common/meta/src/distributed_time_constants.rs